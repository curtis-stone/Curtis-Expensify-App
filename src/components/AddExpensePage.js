import React from "react";
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'

import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component { // exported for testing
  onSubmit = (expense) => {
    // props.dispatch(addExpense(expense));
    this.props.addExpense(expense)
    this.props.history.push('/') // sends back to dashboard onSubmit w/o page refresh (Browser Routing)
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) =>( {addExpense: (expense) => dispatch(addExpense(expense))})
//implicitly return an object
// added to make this component easy to test (mapDisptchTOProps & props.onSubmit...) it merges the props from import component into this compoonent

// onSubmit called when we submit valid data! We will get expense object w/ all propertires
// description, amount, note, and createdAt
// called in props so that ExpenseForm can be reused in other Components
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
// lets component use props too and access to store
