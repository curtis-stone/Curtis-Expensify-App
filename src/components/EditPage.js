import React from "react";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

import ExpenseForm from "./ExpenseForm";

export class EditPage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}> Remove </button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (expense) => dispatch(removeExpense(expense)) // see enzyme docs for props syntax
}); // for tests

const mapStateToProps = (state, props) => {
// props used in argument to get the matching id like above ^
return {
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  )
// find ARRAY method used to search through an array and look for a single item. Returns bool
// returns true if expense id matches match.params.id and object is returned
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditPage);
// undefined in place of mapStateToProps now for testing purposes
