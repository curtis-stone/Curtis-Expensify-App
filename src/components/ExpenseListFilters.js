import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";

import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null, // the dates that we are focusing on
  };
  onDatesChange = ({ startDate, endDate }) => {
    // react-dates library calls this function with an object that has a startDate and endDate
    // destructure these from that object: {startDate, endDate}
    // dispatch the actions to get these to change: setStartDate & setEndDate
    // this.props.dispatch(setStartDate(startDate))
    // this.props.dispatch(setEndDate(endDate))
    // sets start and end dates to the date called by in the react-dates library
    // ^ above is commented out for the dispatch testing purposes below!
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused: calendarFocused }));
    //implicitly sets the state calendarFocus value of null to the dates picked
  };
  onTextChange = (e) => {
    // this.props.dispatch(setTextFilter(e.target.value));
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    // if (e.target.value === "date") {
    //     this.props.dispatch(sortByDate());
    // } else if (e.target.value === "amount") {
    //     this.props.dispatch(sortByAmount());
    // }
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="your_unique_start_date_id"
          endDate={this.props.filters.endDate}
          endDateId="your_unique_end_date_id"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused} // from state above
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

// Value sets the value of our inputs
// we have access to dispatch in components connected to the store!
// use dispatch to log key strokes in input and make changes to the 'text' property in filters
// This component is writing to the store and changing redux store!
// "Controlled Inputs"

const mapStateToProps = (state) => {
  // state gains access to entire state and we return
  // whatever we want
  return {
    filters: state.filters, // in this case we return the state.filters as a property value
    // in an object
  };
}; // can be restructured to return implicitly

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
// connects the store witht he return function and desired target component
