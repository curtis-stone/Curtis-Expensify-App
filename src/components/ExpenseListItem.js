import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ description, amount, createdAt, id }) => ( // already not connected to store so default can be used in testing
  // destructuring used to seperate the properties passed from props.expenses in
  // mapSateToProps in the Expense List Component that this file is contained inside
  // of it's map function
  <div>
    <Link to={`/edit/${id}`}> 
      <h3>{description}</h3>
    </Link>
    <p>
      {" "}
      {amount} - {createdAt}
    </p>
  </div>
);
// onClick calls remove action from actions.expense
// props are put in desctructured object (dispatch and id) b/c they're 
// needed for remove button to work

// template string for LINK
// const mapStateToProps = (state) => {
//   return {
//     expense: state.expense,
//   };
// };

export default ExpenseListItem;

// code has been refactored and some calls to state and mapState/ reduxx store are no longer needed
