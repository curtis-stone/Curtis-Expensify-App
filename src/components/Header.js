import React from 'react';
import {  NavLink } from "react-router-dom";
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth'

export const Header = ({ startLogout }) => ( // exported for testing purposes
    <header>
      <h1>Expensify</h1>
      <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink> 
      <NavLink to="/create" activeClassName="is-active"> Create Expense</NavLink>
      <button onClick={startLogout}> Logout </button>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => { dispatch(startLogout()); }
})
export default connect(undefined, mapDispatchToProps)(Header); 
//mapDispatch.. lets us get startLogout and use it in a way
