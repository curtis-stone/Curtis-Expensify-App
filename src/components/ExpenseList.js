import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => ( // exported for testing w/o store connection
  <div>
    {props.expenses.length === 0 ? (
      <p>No Expenses</p>
    ) : (
      props.expenses.map((expense) => {
        // expenses prop from mapStateToProps, passes down its values to the
        // ExpenseListItem component!
        return <ExpenseListItem key = {expense.id} {...expense} />; // spread operator to get all properties off
        // the expense object
        // don't forget the key prop (should get an error saying this in console)
      })
    )};
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
    // whatever is returned from our selector function in selector's folder will show on screen
    // primarily for filter implementation
  };
};
// this function grabs the state and can return the properties from the state objects
// and make them available in another component's props

export default connect(mapStateToProps)(ExpenseList);
// connect is a react-redux library function that takes a function to grab data from state
// and connect it to another Component. in this case, it makes the data available in props
// for tge Expense List statless functional component!\
// connect is reactive and changes data as state data changes
// export for actual applicaton w/ store connection