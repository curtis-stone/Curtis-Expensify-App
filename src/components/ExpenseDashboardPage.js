import React from "react";

import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseList from "./ExpenseList";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = (props) => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
