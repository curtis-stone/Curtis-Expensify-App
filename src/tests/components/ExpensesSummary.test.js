import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should correctly render expenses summary w/ 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />)
    expect(wrapper).toMatchSnapshot();
})

test('should correctly render expenses summary multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={2356777754} />)
    expect(wrapper).toMatchSnapshot();
})