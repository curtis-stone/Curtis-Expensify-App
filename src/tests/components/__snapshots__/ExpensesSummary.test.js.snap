// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expenses summary multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totaling 
    $23,567,777.54
  </h1>
</div>
`;

exports[`should correctly render expenses summary w/ 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $2.35
  </h1>
</div>
`;
